controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1424,6 +1437,199 @@ func (r *BareMetalHostReconciler) manageHostPower(prov provisioner.Provisioner,
+	}
+
+	// Check if any attach/detach action is pending or failed to attach
+	nodeReservation, nodeLastError := prov.GetDataImageStatus()
+	if nodeReservation != "" {
+		info.log.Info("Node is already under reservation, requeue", "Reserved by", nodeReservation)
+		return actionContinue{dataImageRetryDelay}
+	}
+
+	// Is the current dataImage status valid
+	dirty := false
+	// In case the last node error was not nil for dataimage,
+	// upadate message and counter
+	if nodeLastError != "" {
+		dataImage.Status.Error.Message = nodeLastError
+		dataImage.Status.Error.Count++
+		dirty = true
+	}
+
+	deleteDataImage := false
+	if !dataImage.DeletionTimestamp.IsZero() {
+		deleteDataImage = true
+	}
+
+	requestedURL := dataImage.Spec.URL
+
+	// We can assume non null value since we initialized the status earlier
+	attachedURL := dataImage.Status.AttachedImage.URL
+
+	if deleteDataImage {
+		info.log.Info("DataImage requested for deletion")
+		if attachedURL != "" || dirty {
+			info.log.Info("Detaching DataImage as it was deleted")
+			err := r.detachDataImage(prov, info, dataImage)
+			if err != nil {
+				return actionError{fmt.Errorf("failed to detach, %w", err)}
+			}
+
+			// Requeue to give time to the DataImage Reconciler to update the
+			// status. In case of failure, we will enter this section and
+			// detachDataImage will be called again -> can this cause issues ?
+			return actionContinue{dataImageUpdateDelay}
+		}
+
+		// In case there was no DataImage attached we simply exit
+		// Should we use actionContinue or simply return nil ?
+		return nil
+	}
+
+	if requestedURL != attachedURL || dirty {
+		info.log.Info("DataImage change detected")
+		if attachedURL != "" {
+			info.log.Info("Detaching DataImage")
+			err := r.detachDataImage(prov, info, dataImage)
+			if err != nil {
+				return actionError{fmt.Errorf("failed to detach, %w", err)}
+			}
+
+			// Requeue to give time to the DataImage Reconciler to update the
+			// status. In case of failure, we will enter this section and
+			// detachDataImage will be called again -> can this cause issues ?
+			return actionContinue{dataImageUpdateDelay}
+		}
+		if requestedURL != "" {
+			info.log.Info("Attaching DataImage", "URL", requestedURL)
+			err := r.attachDataImage(prov, info, dataImage)
+			if err != nil {
+				return actionError{fmt.Errorf("failed to attach, %w", err)}
+			}
+
+			// Requeue to give time to the DataImage Reconciler to update the
+			// status. In case of failure, we will enter this section and
+			// attachDataImage will be called again -> can this cause issues ?
+			return actionContinue{dataImageUpdateDelay}
+		}
+	}
+
+	// TODO(hroyrh) : Put a check only if dirty
+	if err := r.Status().Update(info.ctx, dataImage); err != nil {
+		return actionError{fmt.Errorf("failed to update DataImage status, %w", err)}
+	}
+
+	info.log.Info("Updated DataImage Status after handling attachment/detachment")
+
+	// Should we return actionContinue or nil ?
```

Comment:
I think we should return `nil` if nothing changed (no DataImage, already attached/detached), `actionContinue` if we modified something (related to TODO on line 1549).

Reply:
Added a check for `dirty` and returned `actionContinue` on `true`.